cdap-gateway/src/test/java/io/cdap/cdap/gateway/handlers/metrics/MetricsHandlerTest.java

```diff
@@ -780,6 +781,51 @@ public void testResultLimit() throws Exception {
         + (start + 3600), 2, 3);
   }
 
+  @Test
+  public void testDetermineResolution() throws Exception {
+    // if resolution is specified, no matter what the start and end time, it should be that resolution
+    verifyResolution("1s", 0L, 3600L, null, 200, "1s");
+    verifyResolution("1s", 0L, 72000L, null, 200, "1s");
+    verifyResolution("1m", 0L, 100L, null, 200, "60s");
+    verifyResolution("1h", 0L, 100L, null, 200, "3600s");
+
+    // if resolution is auto, the resolution should be based on timestamp, and both timestamp should be provided
+    verifyResolution("auto", null, 300L, Integer.MAX_VALUE, 400, null);
+    // if 0 < ts diff <= 600, second resolution will be used
+    verifyResolution("auto", 0L, 300L, null, 200, "1s");
+    verifyResolution("auto", 0L, 600L, null, 200, "1s");
+    // if 600 < ts diff <= 36000, minute resolution will be used
+    verifyResolution("auto", 0L, 601L, null, 200, "60s");
+    verifyResolution("auto", 0L, 36000L, null, 200, "60s");
+    // if ts diff > 36000, hour resolution will be used
+    verifyResolution("auto", 0L, 36001L, null, 200, "3600s");
+
+    // if resolution is null, the resolution should be same if as resolution is auto if both timestamp are provided
+    verifyResolution(null, 0L, 300L, null, 200, "1s");
+    verifyResolution(null, 0L, 600L, null, 200, "1s");
+    verifyResolution(null, 0L, 601L, null, 200, "60s");
+    verifyResolution(null, 0L, 36000L, null, 200, "60s");
+    verifyResolution(null, 0L, 36001L, null, 200, "3600s");
```

Comment:
this is a great test. It would be good to add cases where the start time is not 0L. Also, what happens if start > end? Should expect a 400?

Reply:
👍 will add more cases